WalletWasabi/Affiliation/AffiliateInformationUpdater.cs

```diff
@@ -0,0 +1,297 @@
+		}
+
+		if (CoinjoinRequests.ContainsKey(roundId))
+		{
+			CoinjoinRequests.Remove(roundId);
+		}
+		else
+		{
+			// This can occurd if the round is finished befor coinjoin requests are updated
+			Logging.Logger.LogInfo($"The round ({roundId}) does not exist.");
+		}
+	}
+
+	private void RemoveRounds()
+	{
+		while (RoundsToRemove.Any())
+		{
+			try
+			{
+				uint256 roundId = RoundsToRemove.Dequeue();
+				RemoveRound(roundId);
+			}
+			catch (Exception exception)
+			{
+				Logging.Logger.LogError(exception.Message);
+			}
+		}
+	}
+
+	private async Task<byte[]?> GetCoinjoinRequestAsync(AffiliateServerHttpApiClient client, Body body)
+	{
+		Payload payload = new(new Header(), body);
+		byte[] signature = Signer.Sign(payload.GetCanonicalSerialization());
+		GetCoinjoinRequestRequest coinjoinRequestRequest = new(body, signature);
+		return await GetCoinjoinRequestAsync(client, coinjoinRequestRequest);
+	}
+
+	private async Task<byte[]?> GetCoinjoinRequestAsync(AffiliateServerHttpApiClient client, GetCoinjoinRequestRequest getCoinjoinRequestRequest)
+	{
+		using CancellationTokenSource cancellationTokenSource = new(AffiliateServerTimeout);
+		CancellationToken cancellationToken = cancellationTokenSource.Token;
+		return await GetCoinjoinRequestAsync(client, getCoinjoinRequestRequest, cancellationToken);
+	}
+
+	private async Task<byte[]?> GetCoinjoinRequestAsync(AffiliateServerHttpApiClient client, GetCoinjoinRequestRequest getCoinjoinRequestRequest, CancellationToken cancellationToken)
+	{
+		try
+		{
+			GetCoinjoinRequestResponse getCoinjoinRequestResponse = await client.GetCoinjoinRequest(getCoinjoinRequestRequest, cancellationToken);
+			return getCoinjoinRequestResponse.CoinjoinRequest;
+		}
+		catch (Exception)
+		{
+			return null;
+		}
+	}
+
+	private void AddAffiliation(uint256 roundId, Coin coin, AffiliationFlag affiliationFlag, bool isPayingZeroCoordinatrionFee)
+	{
+		if (!RoundData.TryGetValue(roundId, out RoundData? roundData))
+		{
+			throw new InvalidOperationException($"The round ({roundId}) does not exist.");
+		}
+
+		roundData.AddInput(coin, affiliationFlag, isPayingZeroCoordinatrionFee);
+	}
+
+	private void AddCoinjoinTransaction(uint256 roundId, NBitcoin.Transaction transaction)
+	{
+		if (!RoundData.TryGetValue(roundId, out RoundData? roundData))
+		{
+			throw new InvalidOperationException($"The round ({roundId}) does not exist.");
+		}
+
+		roundData.AddTransaction(transaction);
+		RoundsToUpdate.Enqueue(roundId);
+	}
+
+	private void CreateRound(uint256 roundId, RoundParameters roundParameters)
+	{
+		RoundData roundData = new();
+		roundData.AddRoundParameters(roundParameters);
+
+		if (!RoundData.TryAdd(roundId, roundData))
+		{
+			throw new InvalidOperationException($"The round ({roundId}) does not exist.");
+		}
+	}
+
+	private void ChangePhase(uint256 roundId, Phase phase)
+	{
+		if (phase == Phase.Ended)
+		{
+			RoundsToRemove.Enqueue(roundId);
```

Comment:
Why to enque and not simply remove it?

Reply:
If I remember it correctly, the only reason was that I wanted to have the event handler as simple as possible because it is called synchronously. Maybe it is an overkill at this place.

Reply:
Yes, imo just removing the round would simplify the code.

Reply:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/cb5525e98eabc24ed5655b2520d1d17642312281.